File: mm-java/src/main/java/io/metamessage/mm/MM.java
package io.metamessage.mm;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * MetaMessage field / type metadata, equivalent to Go struct tag {@code mm:"..."}.
 * Maps to the same binary tag payload as {@code ast.Tag.Bytes()} in the Go implementation.
 */
@Target({ElementType.FIELD, ElementType.TYPE})
@Retention(RetentionPolicy.RUNTIME)
public @interface MM {
    /**
     * When non-empty, overrides the wire key for this field (Go {@code name=} in mm tag).
     * Use {@code "-"} to skip encoding this field (reserved for future use).
     */
    String name() default "";

    ValueType type() default ValueType.UNKNOWN;

    boolean isNull() default false;

    boolean example() default false;

    String desc() default "";

    boolean raw() default false;

    boolean nullable() default false;

    boolean allowEmpty() default false;

    boolean unique() default false;

    /** Go {@code default=}. */
    String defaultValue() default "";

    String min() default "";

    String max() default "";

    int size() default 0;

    /** Pipe-separated enum labels (Go {@code enum=}). */
    String enumValues() default "";

    String pattern() default "";

    /** Timezone offset hours for datetime/date/time; 0 means UTC (omitted on wire when 0). */
    int location() default 0;

    /** UUID / IP wire variant; 0 means default. */
    int version() default 0;

    String mime() default "";

    String childDesc() default "";

    ValueType childType() default ValueType.UNKNOWN;

    boolean childRaw() default false;

    boolean childNullable() default false;

    boolean childAllowEmpty() default false;

    boolean childUnique() default false;

    String childDefault() default "";

    String childMin() default "";

    String childMax() default "";

    int childSize() default 0;

    String childEnum() default "";

    String childPattern() default "";

    int childLocation() default 0;

    int childVersion() default 0;

    String childMime() default "";
}

File: mm-kt/src/main/kotlin/io/github/metamessage/mm/MM.kt
package io.github.metamessage.mm

@Target(AnnotationTarget.FIELD, AnnotationTarget.CLASS)
@Retention(AnnotationRetention.RUNTIME)
annotation class MM(
    val name: String = "",
    val type: ValueType = ValueType.UNKNOWN,
    val isNull: Boolean = false,
    val example: Boolean = false,
    val desc: String = "",
    val raw: Boolean = false,
    val nullable: Boolean = false,
    val allowEmpty: Boolean = false,
    val unique: Boolean = false,
    val defaultValue: String = "",
    val min: String = "",
    val max: String = "",
    val size: Int = 0,
    val enumValues: String = "",
    val pattern: String = "",
    val location: Int = 0,
    val version: Int = 0,
    val mime: String = "",
    val childDesc: String = "",
    val childType: ValueType = ValueType.UNKNOWN,
    val childRaw: Boolean = false,
    val childNullable: Boolean = false,
    val childAllowEmpty: Boolean = false,
    val childUnique: Boolean = false,
    val childDefault: String = "",
    val childMin: String = "",
    val childMax: String = "",
    val childSize: Int = 0,
    val childEnum: String = "",
    val childPattern: String = "",
    val childLocation: Int = 0,
    val childVersion: Int = 0,
    val childMime: String = ""
)
